Reference Documentation issues

- [x] Currently 4.x refdoc generation is failing
- [x] 4.x ref documentation is not listed in the README
- [ ] We shouldn't publish refdocs for SNAPSHOT versions
- [ ] Latest should be the latest released version of the highest major version
- [x] Auto-update the links to latest refdocs for each major version in the README
- [ ] We should check how Spring currently generates their refdocs and see if we need to update

Pasting some investigation notes on the failing 4.x refdoc job:

1) First failure is for `4.6.0` (https://github.com/GoogleCloudPlatform/spring-cloud-gcp/actions/runs/5657280928/job/15592917305).

Suspicious logs:

```
+ for f in ${ROOT_FOLDER}/docs/target/generated-docs/*
+ file='*'
+ copy_docs_for_branch 1.1.x 1.2.x 2.0.0 2.0.0-RC2 2.0.1 2.0.10 2.0.11 2.0.2 2.0.3 2.0.4 2.0.5 2.0.6 2.0.7 2.0.8 2.0.9 3.0.0 3.1.0 3.2.0 3.2.1 3.3.0 3.4.0 3.4.1 3.4.2 3.4.3 3.4.4 3.4.5 3.4.6 3.4.7 3.4.8 3.4.9 3.5.0 3.5.1 3.5.2 3.5.3 3.5.4 3.5.5 3.6.0 3.6.1 3.6.2 4.0.0 4.1.0 4.1.1 4.1.2 4.1.3 4.1.4 4.2.0 4.3.0 4.3.1 4.4.0 4.5.0 4.5.1 4.6.0 Guardfile docs favicon.ico ghpages.sh index.html reference spring-cloud-gcp-autoconfigure spring-cloud-gcp-bigquery spring-cloud-gcp-cloudfoundry spring-cloud-gcp-core spring-cloud-gcp-data-datastore spring-cloud-gcp-data-firestore spring-cloud-gcp-data-spanner spring-cloud-gcp-kms spring-cloud-gcp-logging spring-cloud-gcp-pubsub spring-cloud-gcp-pubsub-stream-binder spring-cloud-gcp-samples spring-cloud-gcp-secretmanager spring-cloud-gcp-security-firebase spring-cloud-gcp-security-iap spring-cloud-gcp-starters spring-cloud-gcp-storage spring-cloud-gcp-vision spring-cloud-generator spring-cloud-previews target ./4.6.0
+ local file=1.1.x
+ local destination=1.2.x
+ git ls-files -i -o --exclude-standard --directory
+ grep -q '^1.1.x$'
+ [[ 1.1.x == \$\{\d\o\c\s\.\m\a\i\n\}\.\h\t\m\l ]]
+ [[ 1.1.x == \s\p\r\i\n\g\-\c\l\o\u\d\-\g\c\p\.\h\t\m\l ]]
+ cp -rf '/home/runner/work/spring-cloud-gcp/spring-cloud-gcp/docs/target/generated-docs/*' 1.2.x
cp: cannot stat '/home/runner/work/spring-cloud-gcp/spring-cloud-gcp/docs/target/generated-docs/*': No such file or directory
```

2) v4.7.0 actually fails for a different reason, during the `mvn install` phase before the updateDocs step:
https://github.com/GoogleCloudPlatform/spring-cloud-gcp/actions/runs/5695458239/job/15569343444
```
Error:  Could not acquire lock(s)
Error:  java.lang.IllegalStateException: Could not acquire lock(s)
Error:  Could not acquire lock(s)
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error: Process completed with exit code 1.
```

3) v4.7.2 fails for the same reason as `4.6.0`, so I expect `v4.7.0`'s failure to be a flake. Interestingly, `v3.6.3` succeeds (released during same cycle as `v4.7.2`), but it uses a slightly different workflow file.

Key differences are in lines 24 and 32. v3.x uses Java 11 and  ` run: echo "::set-output name=date::$(date +'%Y-%m-%d' --utc)"`, whereas v4.x uses Java 17 and `run: echo "date=$(date +'%Y-%m-%d' --utc)" >> $GITHUB_OUTPUT`. 

These differences are minimal and shouldn't be the cause of failure, but just noting here for reference.

---
Something odd is happening with the 4.x workflow where it is copying all versions instead of just the specified one. Will dig into that some more.

--- a/docs/pom.xml
+++ b/docs/pom.xml
@@ -24,11 +24,7 @@
 
 		<!--maven-resources-plugin:3.2.0:copy-resources fails: https://issues.apache.org/jira/browse/MSHARED-966-->
 		<maven-resources-plugin.version>3.3.1</maven-resources-plugin.version>
-		<!-- override version inherited from spring-cloud-build parent to one that is available on Maven Central -->
-		<spring-asciidoctor-backends.version>0.0.4</spring-asciidoctor-backends.version>
 	</properties>
-
-
 
 	<profiles>
 		<profile>

--- a/docs/src/main/asciidoc/firestore.adoc
+++ b/docs/src/main/asciidoc/firestore.adoc
@@ -283,7 +283,7 @@ include::{project-root}/spring-cloud-gcp-data-firestore/src/test/java/com/google
 
 [source, java, indent=0]
 ----
-include::{project-root}/spring-cloud-gcp-data-firestore/src/test/java/com/google/cloud/spring/data/firestore/it/FirestoreIntegrationTestsConfiguration.java[tag=user_service_bean]
+include::{project-root}/spring-cloud-gcp-data-firestore/src/test/java/com/google/cloud/spring/data/firestore/FirestoreIntegrationTestsConfiguration.java[tag=user_service_bean]
 ----
 
 After that, you can autowire your service like so:

--- a/docs/src/main/asciidoc/ghpages.sh
+++ b/docs/src/main/asciidoc/ghpages.sh
@@ -60,12 +60,12 @@ function switch_to_tag() {
 # Build the docs if switch is on
 function build_docs_if_applicable() {
     if [[ "${BUILD}" == "yes" ]] ; then
-        # Disable CI profile (inherited from spring-cloud-build/pom.xml) so
+        # Disable profile CI (inherited from spring-cloud-build/pom.xml) so
         # process-asciidoc can be executed in GH action runner.
         # Profile CI will be activated in GH action runner because
         # there's env.GITHUB_API_URL defined in the environment by
         # default.
-        # See https://github.com/spring-cloud/spring-cloud-build/blob/main/pom.xml#L1406C12-L1419.
+        # See https://github.com/spring-cloud/spring-cloud-build/blob/v4.0.5/pom.xml#L1638-L1652.
         ./mvnw clean install -P docs -P '!CI' -pl docs -DskipTests
     fi
 }

--- a/docs/src/main/asciidoc/ghpages.sh
+++ b/docs/src/main/asciidoc/ghpages.sh
@@ -60,7 +60,7 @@ function switch_to_tag() {
 # Build the docs if switch is on
 function build_docs_if_applicable() {
     if [[ "${BUILD}" == "yes" ]] ; then
-        ./mvnw clean install -P docs -pl docs -DskipTests -q
+        ./mvnw clean install -P docs -pl docs -DskipTests
     fi
 }
 

--- a/docs/src/main/asciidoc/ghpages.sh
+++ b/docs/src/main/asciidoc/ghpages.sh
@@ -60,7 +60,13 @@ function switch_to_tag() {
 # Build the docs if switch is on
 function build_docs_if_applicable() {
     if [[ "${BUILD}" == "yes" ]] ; then
-        ./mvnw clean install -P docs -pl docs -DskipTests
+        # Disable CI profile (inherited from spring-cloud-build/pom.xml) so
+        # process-asciidoc can be executed in GH action runner.
+        # Profile CI will be activated in GH action runner because
+        # there's env.GITHUB_API_URL defined in the environment by
+        # default.
+        # See https://github.com/spring-cloud/spring-cloud-build/blob/main/pom.xml#L1406C12-L1419.
+        ./mvnw clean install -P docs -P '!CI' -pl docs -DskipTests
     fi
 }
 

